### ISSUE TITLE
Can you add support for floorp browser?

### ISSUE THREAD

**Issue by mobsenpai** [nix-community/home-manager#4912]
### Description
![Floorp](https://github.com/Floorp-Projects/Floorp)

Currently If I use `package = pkgs.floorp` in my firefox settings configured using home-manager, it does not update the settings for floorp browser. I checked and there is no module for it atm. If I am not wrong a seperate module is needed for it to work as floorp browser does not use `~/.mozilla` folder for configs instead it uses `~/.floorp`. I can't use it due to this reason but want to, please help. Please correct me if I am doing anything wrong.

**Comment by sbatial** [comment]
I came to a similar conclusion but as far as I can tell a new option setting the config location would be better (in my opinion) because that seems to be the only differentiating factor from Firefox.

I can't say whether I find the time but I'll try to get it to work.

Just a thought... But have you tried sym-linking `.floorp` to `.mozilla`? (even just from the command-line without using home-manager.
If that works you should be able to use hm's link option)
That should
1. be a hotfix until an official way exists and
2. indicate if that's actually all that is needed

**Comment by MobSenpai** [comment]
Wait! that's honestly a good idea. I will try to symlink tonight, If I get time. 

> I came to a similar conclusion but as far as I can tell a new option setting the config location would be better (in my opinion) because that seems to be the only differentiating factor from Firefox.
> 
> I can't say whether I find the time but I'll try to get it to work.
> 
> Just a thought... But have you tried sym-linking `.floorp` to `.mozilla`? (even just from the command-line without using home-manager. If that works you should be able to use hm's link option) That should
> 
>     1. be a hotfix until an official way exists and
> 
>     2. indicate if that's actually all that is needed

**Comment by xopclabs** [comment]
I've been able to make it work in impure mode by adding this:
```nix
home.file.".mozilla/firefox/profileName/chrome" = {
    recursive = true;
    source = ./firefox-nordic-theme;
};
home.file.".floorp/profileName.default" = {
    recursive = true;
    source = "/home/username/.mozilla/firefox/profileName";
};
home.file.".floorp/profiles.ini".text = ''
    [Profile0]
    Name=default
    IsRelative=1
    Path=profileName.default
    Default=1

    [General]
    StartWithLastProfile=1
    Version=2
'';
```
Even custom css I'm adding with home-manager to a firefox directory is working! It's hacky but it works for now. Maybe I'll look into creating a hm module

**Comment by MobSenpai** [comment]
> I've been able to make it work in impure mode by adding this:
> 
> ```nix
> home.file.".floorp/profileName.default" = {
>     recursive = true;
>     source = "/home/username/.mozilla/firefox/profileName";
> };
> home.file.".floorp/profiles.ini".text = ''
>     [Profile0]
>     Name=default
>     IsRelative=1
>     Path=profileName.default
>     Default=1
> 
>     [General]
>     StartWithLastProfile=1
>     Version=2
> '';
> ```
> 
> Even extensions and custom css I'm adding with home-manager to a firefox directory is working! It's hacky but it works for now. Maybe I'll look into creating a hm module
> 
> The only difference between `.floorp` and `.mozilla/firefox` is that floorp has `Version=2` under `General` block in `profiles.ini`. Haven't looked into what it means, though!

Hmm this definitely is better hotfix. Also its good news that extensions are working cuz I had no idea why it didn't. Would love a module soon. No pressure tho. Also thanks everyone for investing time on it.

**Comment by xopclabs** [comment]
Hey! I've added a floorp module to HM and it seems to be working fine! I need to look into submitting a PR, but I think it would need a lot more work to be approved (and since I know very little of nix language tbh, It can take a while)
In the mean time, you can try to use my fork of home-manager. Replace `"github:nix-community/home-manager"` in your `flake.nix` with `"github:xopclabs/home-manager/floorp-browser"` and you can use `programs.floorp` as a drop-in replacement for `programs.firefox` in your home-manager config

**Comment by MobSenpai** [comment]
> Hey! I've added a floorp module to HM and it seems to be working fine! I need to look into submitting a PR, but I think it would need a lot more work to be approved (and since I know very little of nix language tbh, It can take a while) In the mean time, you can try to use my fork of home-manager. Replace `"github:nix-community/home-manager"` in your `flake.nix` with `"github:xopclabs/home-manager/floorp-browser"` and you can use `programs.floorp` as a drop-in replacement for `programs.firefox` in your home-manager config

Wow! Thank you so much @xopclabs. Give the PR link when you submit it. I should close this issue with that isn't it?

**Comment by xopclabs** [comment]
Yeah, I guess! I've noticed that #4926 does almost exactly the same but for Mullvad Browser, so I'll look into it

**Comment by sbatial** [comment]
@xopclabs Makes me wonder if it would make sense to combine some of those configs into a shared modules to cut down on a lot of duplicate config for different firefox derivatives.

I'm also no nix expert by any means but I'll try to have a look at the firefox, mullvad and your floorp module when I find the time.

(I also don't know if there is any precedent for such a thing (de-duplication vs. standalone modules) in the HM codebase but that would be a seperate issue anyways)

**Comment by brckd** [comment]
`programs.librewolf` could also make use of this, as it currently doesnt have all the options that `programs.firefox` provides.

**Comment by sbatial** [comment]
Yea, so I had a quick look and as far as I can tell the floorp module is almost the same as the firefox module (which makes sense haha) apart from the name.
So if desired it should be totally possible to merge those configs.

(I have too much on my plate right now for it though :raised_hands: and am currently not using Floorp anyways)

**Comment by brckd** [comment]
I have a day off today, so i'll try to look into it too! It's a roadblock on my nix config anyways as i don't want to use stock Firefox.

**Comment by brckd** [comment]
So as you mentioned basically there were two solutions to this.

1. Create a separate new module for each Firefox derivative, like done in #4926
Pros: This allows for different configuration options.
3. Create one module for all Firefox and change the config path depending on the provided `package`
Pros: Less repetitive code.

However, the same problem has already been solved for Chromium in #277. Take a look at this code snippet: https://github.com/nix-community/home-manager/blob/cf111d1a849ddfc38e9155be029519b0e2329615/modules/programs/chromium.nix#L205-L217
Basically, there is one `chromium.nix` module that creates options and configs for all the other Chromium derivatives. The same could be done in `firefox.nix` with its own `browserModule` method. These options and configs could be overridden in place or in separate modules like `floorp.nix`.

Not sure if it's good practice to define options and configs in separate modules like `firefox.nix` and `floorp.nix`, but I'll create a PR with the changes soon.

**Comment by brckd** [comment]
I tried implementing the ˋbrowserModuleˋ method [here](https://github.com/brckd/home-manager/blob/92ea45beb1bd4b6e5fe98e277a450aadcf4b7b60/modules/programs/firefox.nix#L939-L955), but couldn't get it to work with anything other than Firefox.

**Comment by brckd** [comment]
I am able to configure Floorp similar to `programs.firefox` on [this branch](https://github.com/brckd/home-manager/tree/ff-forks/floorp).

**Comment by mobsenpai** [comment]
Not using floorp anymore. So closing it.

**Comment by mobsenpai** [comment]
Not using floorp anymore. Closing it.

**Comment by mobsenpai** [comment]
Not using floorp anymore.

**Comment by kbwhodat** [comment]
For anyone wanting Floorp to work on both nix-darwin and nixos with your userChrome.css, user.js, extensions etc applied this is a module I referenced from an individual:
https://github.com/kbwhodat/configs/blob/main/nix-config/modules/floorp.nix

From there you can reference the module:
https://github.com/kbwhodat/configs/blob/main/nix-config/common/browsers/floorp.nix

For Darwin, you have to use an [overlay that provides the binary package from official builds](https://github.com/kbwhodat/nixpkgs-firefox-darwin). I created a fork for adding Floorp, waiting for the PR to be merged. Here's how you would add the input to your flake.nix.
https://github.com/kbwhodat/configs/blob/main/nix-config/flake.nix#L18

Hope this helps somebody
